File: backend/api/services/ai_service.py
# ==========================================================
# services/ai_service.py — OpenAI呼び出しのラッパー。
# プロンプト文自体は services/prompt.py が組み立てる。
# ここではAPIキー未設定時のガード、実際のAPI呼び出し、
# JSON応答のパース・不足キーの補完だけを担当する。
# ==========================================================

import json
from openai import OpenAI
from fastapi import HTTPException
import traceback

import config
from services.prompt import build_prompt


def generate_ai_note(selected_text: str) -> dict:

    if not config.OPENAI_API_KEY:
        # AI解説は任意機能。キー未設定時にモジュール読み込み時点でクラッシュしないよう、
        # クライアントはここで初めて生成し、事前にわかりやすいエラーを返す。
        # 呼び出し側（拡張機能）はこの失敗を無視してマーカー機能自体は継続する。
        raise HTTPException(
            status_code=503,
            detail="OPENAI_API_KEYが未設定のため、AI解説は生成されません。",
        )

    client = OpenAI(api_key=config.OPENAI_API_KEY)

    prompt = build_prompt(selected_text)

    try:
        response = client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[
                {
                    "role": "system",
                    "content": (
                        "あなたは語句の解説をするアシスタントです。"
                        "必ずJSON形式のみで返してください。"
                        "余計な文字列は一切含めないでください。"
                    )
                },
                {
                    "role": "user",
                    "content": prompt
                }
            ],
            temperature=0.3,
            max_tokens=500,
            response_format={"type": "json_object"},
        )

        raw = response.choices[0].message.content
        result = json.loads(raw)

        required_keys = ["explanation", "similar_words", "antonyms", "translation", "usage_example"]
        for key in required_keys:
            if key not in result:
                result[key] = ""

        return result

    except Exception as e:
        # 詳細なエラーをログに出力
        print("=== AI SERVICE ERROR ===")
        print(f"Type: {type(e).__name__}")
        print(f"Message: {str(e)}")
        traceback.print_exc()
        print("========================")
        raise HTTPException(
            status_code=500,
            detail=f"{type(e).__name__}: {str(e)}"
        )

File: extension/modules/aiService.js
// ==========================================================
// modules/aiService.js — OpenAIのChat Completions APIを直接叩く。
// 旧backend/api/services/{ai_service,prompt}.pyの置き換え。
// APIキーはui/settings.htmlで入力してもらい、dataStore.js経由でchrome.storage.localに置く。
// ==========================================================

function buildPrompt(selectedText) {
  return `以下の語句または文章について、日本語で次の5項目を答えてください。
出力はJSONのみで、前置きや説明は書かないでください。

対象: 「${selectedText}」

{
  "explanation": "意味・解説（2〜4文）",
  "similar_words": "類似語・同義語（カンマ区切りで3〜5語）",
  "antonyms": "対義語（カンマ区切りで1〜3語、なければ「なし」）",
  "translation": "日本語訳（対象がすでに日本語なら「なし」）",
  "usage_example": "例文（自然な日本語で1文）"
}`;
}

async function generateAiNoteViaOpenAi(selectedText) {
  const settings = await dsGetSettings();
  const apiKey = (settings.openaiApiKey || "").trim();

  if (!apiKey) {
    throw new Error("OpenAIのAPIキーが設定されていません。拡張機能の設定画面から登録してください。");
  }

  let response;
  try {
    response = await fetch("https://api.openai.com/v1/chat/completions", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify({
        model: "gpt-4o-mini",
        temperature: 0.3,
        max_tokens: 500,
        response_format: { type: "json_object" },
        messages: [
          { role: "system", content: "語句の解説を行うアシスタントです。JSONのみで回答してください。" },
          { role: "user", content: buildPrompt(selectedText) },
        ],
      }),
    });
  } catch (error) {
    throw new Error(`OpenAIに接続できませんでした: ${error.message}`);
  }

  if (!response.ok) {
    const body = await response.text().catch(() => "");
    let message = `OpenAI APIエラー (${response.status})`;
    try {
      message = JSON.parse(body).error.message;
    } catch {
      // JSONで返ってこなければ上のステータスコードのみのメッセージを使う
    }
    throw new Error(message);
  }

  const data = await response.json();
  let note;
  try {
    note = JSON.parse(data.choices[0].message.content);
  } catch {
    throw new Error("AIの応答を解析できませんでした");
  }

  for (const key of ["explanation", "similar_words", "antonyms", "translation", "usage_example"]) {
    if (!(key in note)) note[key] = "";
  }
  return note;
}
